experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,654 @@
+        mapp = hom(mixed_ring, group_ring, vcat([0 for i in 1:m], gens(group_ring), [0 for i in 1:n]))
+        Mat = matrix(group_ring, nrows(big_matrix), nrows(big_matrix), [mapp(big_matrix[i,j]) for i in 1:nrows(big_matrix), j in 1:nrows(big_matrix)])
+        return Mat
+    else
+        l = length(v)
+        #there are l*m^2 z's
+        ringg,Z = PolynomialRing(QQ,"Z"=>1:l*m^2)
+        prod_mat = matrix(ringg,1,1,[1])
+        for i in 1:l
+            mat = rep_mat_(m,v[i])
+            R = base_ring(mat)
+            mapp = hom(R,ringg, gens(ringg)[((i-1)*m^2)+1:i*m^2])
+            mat_ = matrix(ringg, nrows(mat), ncols(mat), [0 for i in 1:nrows(mat)*ncols(mat)])
+            for i in 1:nrows(mat)
+                for j in 1:ncols(mat)
+                    mat_[i,j] = mapp(mat[i,j])
+                end
+            end
+            prod_mat = kronecker_product(prod_mat, mat_)
+        end
+        return prod_mat
+    end
+end
+
+function direct_sum(M::AbstractAlgebra.Generic.MatSpaceElem, N::AbstractAlgebra.Generic.MatSpaceElem)
+    #parent(M[1,1]) == parent(N[1,1]) || @error("not same ring")
+    mr = nrows(M)
+    mc = ncols(M)
+    nr = nrows(N)
+    nc = ncols(N)
+    mat_ = matrix(base_ring(M), mr+nr, mc+nc, [0 for i in 1:(mr+nr)*(mc+nc)])
+    mat_[1:mr, 1:mc] = M
+    mat_[mr+1:mr+nr, mc+1:mc+nc] = N
+    return mat_
+end
+
+function num_of_as(m::Int, v::Vector{Int})
+    max = maximum(v)
+    sum_ = binomial(m + max - 1, m - 1)
+    return sum_
+end
+
+##########################
+#Invariant Rings of Reductive groups
+##########################
+mutable struct InvariantRing
+    field::Field
+    poly_ring::MPolyDecRing #graded
+    
+    group::ReductiveGroup
+    representation::RepresentationReductiveGroup
+    
+    fundamental::Vector{MPolyDecRingElem}
+    
+    reynolds_operator::Function
+    
+    NullConeIdeal::MPolyIdeal
+    
+    #This should be the only one that is used
+    function InvariantRing(R::RepresentationReductiveGroup) #here G already contains information n and rep_mat
+        z = new()
+        n = ncols(R.rep_mat)
+        z.representation = R
+        z.group = R.group
+        G = z.group
+        z.field = G.field
+        z.poly_ring, __ = grade(PolynomialRing(G.field, "X" => 1:n)[1])
+        z.reynolds_operator = reynolds_v_slm
+        I, M = proj_of_image_ideal(G, R.rep_mat)
+        z.NullConeIdeal = ideal(generators(G, I, R.rep_mat))
+        z.fundamental = inv_generators(z.NullConeIdeal, G, z.poly_ring, M, I, z.reynolds_operator)
+        return z
+    end
+    
+    function InvariantRing(R::RepresentationReductiveGroup, ring::MPolyRing)
+        n = ncols(R.rep_mat)
+        n == ngens(ring) || error("The given polynomial ring is not compatible.")
+        z = new()
+        z.representation = R
+        z.group = R.group
+        G = R.group
+        z.field = G.field
+        if is_graded(ring)
+            z.poly_ring = ring
+        else
+            z.poly_ring = grade(ring)
```

Comment:
Really? Should it perhaps be
```suggestion
            z.poly_ring = grade(ring)[1]
```
It would be good if you edited the `runtests.jl` file added by this PR, and inserted some `@test` invocations in here, to test the code. Then one of the tests could be done with both graded and ungraded inputs, to trigger both code paths above. That would ensure it works, and that it keeps working.

Please have a look at the files inside the global `test/` directory, as well as those in other `experimental/*/test` directories to get an idea of how those look like. The Julia documentation  also talks about tests: https://docs.julialang.org/en/v1/stdlib/Test/

Also read <https://docs.oscar-system.org/dev/Experimental/intro/#Useful-functions-for-development>


And while I am at it, perhaps read our dev docs:
- https://docs.oscar-system.org/dev/DeveloperDocumentation/new_developers/
- https://docs.oscar-system.org/dev/DeveloperDocumentation/styleguide/
- and so on